components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/WebExtensionToolbarFeature.kt

```diff
@@ -34,45 +38,66 @@ class WebExtensionToolbarFeature(
+        iconThread.start()
+        Handler(iconThread.looper)
+    }
+
+    internal lateinit var iconJobDispatcher: CoroutineDispatcher
+
     /**
      * Starts observing for the state of web extensions changes
      */
     override fun start() {
+        iconJobDispatcher = iconHandler.asCoroutineDispatcher("WebExtensionIconDispatcher")
+
         scope = store.flowScoped { flow ->
             flow.ifAnyChanged { arrayOf(it.selectedTab, it.extensions) }
                 .collect { state ->
-                    state.selectedTab?.let { tab ->
-                        renderWebExtensionActions(state, tab)
-                    }
+                    renderWebExtensionActions(state, state.selectedTab)
                 }
         }
     }
 
     override fun stop() {
+        iconJobDispatcher.cancel()
         scope?.cancel()
     }
 
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
-    internal fun renderWebExtensionActions(state: BrowserState, tab: SessionState) {
-        val extensionsMap = state.extensions.toMutableMap()
-        extensionsMap.putAll(tab.extensionState)
-
-        val extensions = extensionsMap.values.toList()
-
+    internal fun renderWebExtensionActions(state: BrowserState, tab: SessionState?) {
+        val extensions = state.extensions.values.toList()
         extensions.forEach { extension ->
-            extension.browserAction?.let { extensionAction ->
-                val existingBrowserAction = webExtensionBrowserActions[extension.id]
-
-                if (existingBrowserAction != null) {
-                    existingBrowserAction.browserAction = extensionAction
-                } else {
+            extension.browserAction?.let { browserAction ->
+                // Add the global browser action if it doesn't exist
+                val toolbarAction = webExtensionBrowserActions.getOrPut(extension.id) {
                     val toolbarAction = WebExtensionToolbarAction(
-                            browserAction = extensionAction,
-                            listener = extensionAction.onClick)
+                        browserAction = browserAction,
+                        listener = browserAction.onClick,
+                        iconJobDispatcher = iconJobDispatcher
+                    )
                     toolbar.addBrowserAction(toolbarAction)
-                    webExtensionBrowserActions[extension.id] = toolbarAction
+                    toolbar.invalidateActions()
+                    toolbarAction
+                }
+
+                // Check if we have a tab-specific override and apply it.
+                // Note that tab-specific actions only contain the values that
+                // should be overridden. All other values will be null and should
+                // therefore *not* be applied / overridden.
+                tab?.extensionState?.get(extension.id)?.browserAction?.let {
+                    toolbarAction.browserAction = BrowserAction(
```

Comment:
This we can't do because both the tab-specific override and the global action can have null values and we need to be able to differentiate between say a null or empty badgeText.